query/parse.go

```diff
@@ -72,11 +72,21 @@ func isSpace(c byte) bool {
 	return c == ' ' || c == '\t'
 }
 
+// globally tracking parentheses
+var parenBalance int
+
 // Parse parses a string into a query.
 func Parse(qStr string) (Q, error) {
 	b := []byte(qStr)
+	parenBalance = 0
+	qs, _, err := parseExprList(b, &parenBalance)
 
-	qs, _, err := parseExprList(b)
+	if parenBalance < 0 {
+		return nil, fmt.Errorf("right parentheses unbalanced")
+	}
+	if parenBalance > 0 {
+		return nil, fmt.Errorf("left parentheses unbalanced")
+	}
 	if err != nil {
```

Comment:
prefer returning the err before checking parenBalance. If err is set it is likely more useful than knowing paren balance